tools/libcSpecGen.cc

```diff
@@ -0,0 +1,511 @@
+#include <vector>
+#include <iostream>
+#include <sys/stat.h>
+#define numOps 10
+
+#define sizeof_array(ARRAY) (sizeof(ARRAY) / sizeof(ARRAY[0]))
+
+const std::string mkread(const std::string &var)
+{
+  return std::string("sea_dsa_set_read(" + var + ");\n");
+}
+
+const std::string mkwrite(const std::string &var)
+{
+  return std::string("sea_dsa_set_modified(" + var + ");\n");
+}
+
+const std::string mkheap(const std::string &var)
+{
+  return std::string("sea_dsa_set_heap(" + var + ");\n");
+}
+
+std::string collapse(const std::string &var)
+{
+  return std::string("sea_dsa_collapse(" + var + ");\n");
+}
+
+std::string alias(const std::vector<std::string> &vect)
+{
+  std::string callsite;
+  callsite = "sea_dsa_alias(";
+
+  for (int i = 0; i < vect.size() - 1; i++)
+  {
+    callsite.append(vect[i]);
+    callsite.append(", ");
+  }
+  callsite.append(vect[vect.size() - 1]);
+  callsite.append(");\n");
+
+  return callsite;
+}
+
+struct libAction
+{
+  // The return value/arguments that should be marked read.
+  bool read[numOps];
+
+  // The return value/arguments that should be marked modified.
+  bool write[numOps];
+
+  // The return value/arguments that should be marked as heap.
+  bool heap[numOps];
+
+  // Flags whether the return value should be merged with all arguments.
+  bool mergeNodes[numOps];
+
+  // Flags whether the return value and arguments should be folded.
+  bool collapse;
+};
+
+#define NRET_NARGS               \
+  {                              \
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0 \
+  }
+#define YRET_NARGS               \
+  {                              \
+    1, 0, 0, 0, 0, 0, 0, 0, 0, 0 \
+  }
+#define NRET_YARGS               \
+  {                              \
+    0, 1, 1, 1, 1, 1, 1, 1, 1, 1 \
+  }
+#define YRET_YARGS               \
+  {                              \
+    1, 1, 1, 1, 1, 1, 1, 1, 1, 1 \
+  }
+#define NRET_NYARGS              \
+  {                              \
+    0, 0, 1, 1, 1, 1, 1, 1, 1, 1 \
+  }
+#define YRET_NYARGS              \
+  {                              \
+    1, 0, 1, 1, 1, 1, 1, 1, 1, 1 \
+  }
+#define NRET_YNARGS              \
+  {                              \
+    0, 1, 0, 0, 0, 0, 0, 0, 0, 0 \
+  }
+#define YRET_YNARGS              \
+  {                              \
+    1, 1, 0, 0, 0, 0, 0, 0, 0, 0 \
+  }
+#define YRET_NNYARGS             \
+  {                              \
+    1, 0, 0, 1, 1, 1, 1, 1, 1, 1 \
+  }
+#define YRET_YNYARGS             \
+  {                              \
+    1, 1, 0, 1, 1, 1, 1, 1, 1, 1 \
+  }
+#define NRET_NNYARGS             \
+  {                              \
+    0, 0, 0, 1, 1, 1, 1, 1, 1, 1 \
+  }
+#define YRET_NNYNARGS            \
+  {                              \
+    1, 0, 0, 1, 0, 0, 0, 0, 0, 0 \
+  }
+#define NRET_NNNYARGS            \
+  {                              \
+    0, 0, 0, 0, 1, 1, 1, 1, 1, 1 \
+  }
+
+const struct
+{
+  const std::string name;
+  const libAction action;
+} recFuncs[] = {
```

Comment:
acknowledge where this table is from

Reply:
would a link to the repo / file in repo suffice?